edgraph/server.go

```diff
@@ -668,6 +674,56 @@ func (s *Server) doMutate(ctx context.Context, qc *queryContext, resp *api.Respo
 	return nil
 }
 
+// validateMutation ensures that the value in the edge is not too big.
+// The challange here is that the keys in badger have a limitation on their size (< 2<<16).
+// We need to ensure that no key, either primary or secondary index key is bigger than that.
+// See here for more details: https://github.com/dgraph-io/projects/issues/73
+func validateMutation(ctx context.Context, edges []*pb.DirectedEdge) error {
+	// key = meta data + predicate + actual key, this all needs to fit into 64 KB
+	// we are keeping 1 KB aside for meta information we put into the key and we
+	// use 63 KB for the rest, that is predicate and the actual key.
+	const maxKeySize = 63 << 10
+
+	for _, e := range edges {
+		maxSizeForDataKey := maxKeySize - len(e.Attr)
+
+		// seems reasonable to assume, the tokens for indexes won't be bigger than the value itself
+		if len(e.Value) <= maxSizeForDataKey {
+			continue
+		}
+		pred := x.NamespaceAttr(e.Namespace, e.Attr)
+		update, ok := schema.State().Get(ctx, pred)
+		if !ok {
+			continue
+		}
+		// only string type can have large values that could cause us issues later
+		if update.GetValueType() != pb.Posting_STRING {
+			continue
+		}
+
+		storageVal := types.Val{Tid: types.TypeID(e.GetValueType()), Value: e.GetValue()}
+		schemaVal, err := types.Convert(storageVal, types.TypeID(update.GetValueType()))
+		if err != nil {
+			return fmt.Errorf("too large value to store: %w", err)
```

Comment:
I think this error could be misleading, there are many reasons that types.Convert might fail. Maybe better to errors.Wrap with a more generic message?

Reply:
Done